Adjust regtest to new config locations

mgerstner opened the pull request:
An alternative approach to PR #49 as discussed between maltek and me.

mgerstner committed:
chkstat: suppport new --config-root command line option

This new switch allows to read configuration files from an alternative
root directory. Together with `--root` this allows to operate completely
on a sub-directory tree.

This option will make it easier to correctly implement tests/regtest.py
now that permission profiles are located under /usr/share/permissions by
default.

mgerstner committed:
regtest: move initialialization of TestBase paths into the prepare() function

Make related paths class scope variables. This allows them to be based
on information retrieved from the main test class. And they don't need
to be reset for each new test instance.

Also fix the `global_init_performed` logic which always triggered
before.

Also make the `chkstat` copy available in PATH.

mgerstner committed:
regtest: mount permissions src repository in /usr/src/permissions

This is a more natural place for it and danger is hopefully lower that
is will be accidentally removed from within the test environment.

mgerstner committed:
regtest: provide valid home directory in /root

mgerstner committed:
regtest: only catch explicit FileNotFoundError

mgerstner committed:
regtest: adjust test suite to new path locations in /usr/share/permissions

Since chkstat now expects the standard profiles to reside in
/usr/share/permissions and only accepts /etc/permissions.* as a legacy
fallback, the test suite needs to cope with this new situation.

The current approach of recursively copying /etc from the host file
system can no longer be used for /usr/share, since it may contain
gigabytes of data. An overlayfs mount in /usr/share combining a tmpfs
from the unprivileged namespace and the static data from the host file
system would be the perfect solution. Alas, for security reasons
overlayfs mounts are not allowed un unprivileged namespaces.

Another approach shown in PR#49 simply bind-mounts everything in /etc/*,
/usr/* and /usr/share/* resulting in hundreds of bind mounts in the test
environment. This also didn't seem ideal.

Therefore this commit uses the new `--config-root` feature in chkstat to
point it to an alternate file system location for looking up config
files. This way also /etc now longer needs to be copied from the host
filesystem. Instead beneath /usr/local a tmpfs with all necessary
configuration ingredients is setup. The commit also needs to adjust some
details since the local profile is now in a separate path (still in in
/etc) while the default profiles are found in the new path
(/usr/share/permissions).

maltek commented on tests/regtest.py:
```diff
@@ -784,11 +780,26 @@ def getDescription(self):
+			# make sure base dirs exist
+			os.makedirs(self.s_sysconfig, 0o755, exist_ok = True)
+			os.makedirs(self.s_permissions_dir, 0o755, exist_ok = True)
+			os.makedirs(self.s_profile_root, 0o755, exist_ok = True)
```
Maybe move this into `createTestFile`? That'd already allow getting rid of two of these variables.

mgerstner replied:
Hmm I'm not quite sure what you mean yet. `createTestfile` currently is used to create test case specific files somewhere in `/TestCaseDir/testfile`, while these three here create "system configuration directories" that should only be created once and not disappear at any later time.

maltek replied:
The idea is to do this `mkdir -p` implicitly when the config files are written, so that we can get rid of two of the variables that make it hard to figure out what each path variable refers to. Maybe that's the wrong function, I just saw it being used to create the permissions config files in `resetConfigs`.

mgerstner commented on tests/regtest.py:
```diff
@@ -486,6 +486,9 @@ def checkCapsSupport(self):
 
+	def getChkstatConfigRoot(self):
+		return "/usr/local"
+
```
I deliberately made them methods to allow potentially more complex logic in the future. Judging from the evolution this whole test setup made by now it's quite conceivable that additional complexity happens ;)

maltek replied:
I'm usually not a fan of this kind of proactive complexity, because way too often later developments are not perfectly predicted so you just end up with an even more complex result in the end. (The buzzword for this is YAGNI.)

This is just a simple function here though, so there's not much point in bikeshedding this too much.

mgerstner commented on tests/regtest.py:
```diff
@@ -791,15 +782,23 @@ def getDescription(self):
+			# make a convenience symlink to make it feel more
+			# natural in /usr/local
+			os.symlink(config_root + "/usr/share", config_root + "/share")
 
```
I know what you mean. What do you suggest? Better variable names or redundant literal strings? I'd hate to fix the paths throughout the code.

maltek replied:
https://github.com/openSUSE/permissions/pull/51#discussion_r390347578 is one suggestion

Probably that's not enough - then I guess redundant string literals are the way to go.

mgerstner commented on src/chkstat.cpp:
```diff
@@ -439,41 +448,37 @@ collect_permfiles()
 
-        snprintf(fn, sizeof(fn), "/etc/permissions.%s", level[i]);
-        if (access(fn, R_OK) == 0)
+        std::string profile = (etc_root + "/permissions.") + level[i];
```
No they're not required. Must be a remnant of some other version of the line I had there before. I'll remove them.

mgerstner commented:
I force-pushed a version without the superfluous parantheses.

mgerstner committed:
regtest: improve readability of path variables by using literals

It's hard to keep track of the various path variables so in favour of
readability this commit changes to a literal path suffix approach. Only
the config_root remains centrally assigned.

mgerstner commented on tests/regtest.py:
```diff
@@ -842,9 +841,16 @@ def addProfileEntries(self, entries):
 
+		def getAgnosticProfilePath(profile):
+
+			if profile == "local":
```
Yeah I fixed it.

mgerstner commented:
I've added a commit on top that changes the approach of the path variables to use literal path suffixes. I think this makes it easier readable.

maltek commented:
Seems that we did a good job talking past each other. I didn't actually (mean to) ask to remove the variables altogether, just to define the variables using string literals as opposed to defining them using other variables.

I'm fine with that radical solution as well, though. You decide if you want to do another pass or not.

mgerstner commented:
> Seems that we did a good job talking past each other. I didn't actually (mean to) ask to remove the variables altogether, just to define the variables using string literals as opposed to defining them using other variables.

Okay then I didn't fully grasp what you meant.

> I'm fine with that radical solution as well, though. You decide if you want to do another pass or not.

I think the current solution is also fine. I like it better actually. It's more clear what paths are operated on without having to learn the semantics of variable names or encode paths in variable names.
